src/migrations/20240520000001-create-timeseries-and-root-causes.js

```diff
@@ -0,0 +1,352 @@
+const {
+  prepMigration,
+} = require('../lib/migration');
+
+/** @type {import('sequelize-cli').Migration} */
+module.exports = {
+  async up(queryInterface) {
+    await queryInterface.sequelize.transaction(async (transaction) => {
+      const sessionSig = __filename;
+      await prepMigration(queryInterface, transaction, sessionSig);
+
+      // This starts by creating a stored procedure that can be
+      // used now and in the future to create a time series table with a
+      // name of the form <source tablename>_timeseries, e.g. Goals_timeseries
+
+      // usage: SELECT create_timeseries_from_audit_log('<tablename>')
+      // example: SELECT create_timeseries_from_audit_log('Goals')
+      // creates Goals_timeseries
+
+      // Resulting tables are like their source tables except:
+      // * id becomes data_id and is only unique when combined with the timeband
+      // * timeband_start identifies the beginning of the timeband of a particular
+      //   record state
+      // * timeband_end identifies the ending of the timeband were the record has
+      //   that state
+      // * enums are converted to text values because enums change over time and
+      //   thus historical data may not fit current enums
+
+      // Root causes were not saving to ActivityReportGoalFieldResponses
+      // on multi-recipient ARs where a goal already had a root cause
+      // for one recipient. Because ActivityReportGoalFieldResponses
+      // is supposed to have the root cause as it was at report approval
+      // time, which can differ from the current root causes for goals
+      // as recorded in GoalFieldResponses. Calculating what the state of
+      // root causes at the time of AR approval is aided by having a
+      // time series of GoalFieldResponses to join into with the addition of
+      // of a check like:
+      // LEFT JOIN GoalFieldResponses_timeseries gfrt
+      //   ON arg."goalId" = gfrt."goalId"
+      //   AND ar."approvedAt" BETWEEN gfrt.timeband_start AND gfrt.timeband_end
+
+      await queryInterface.sequelize.query(`
+
+      CREATE OR REPLACE FUNCTION create_timeseries_from_audit_log(tablename text)
+      RETURNS VOID LANGUAGE plpgsql AS
+      $$
+      DECLARE
+        qry text := '';
+        wtext text := '';
+        rec record;
+      BEGIN
+      -- Get the column list for the main table
+      qry := format('
+      DROP TABLE IF EXISTS clist;
+      CREATE TEMP TABLE clist
+      AS
+      SELECT
+        column_name cname
+        ,ordinal_position cnum
+        ,data_type  ctype
+      FROM information_schema.columns ic
+      WHERE table_schema = %L
```

Comment:
Finding my self scratching my head on some of this stuff. Can we add some comments on some of the more cryptic lines. Even a 1 line comment would be helpful.